x-pack/plugins/cloud_security_posture/server/routes/graph/v1.ts

```diff
@@ -36,50 +40,69 @@ interface LabelEdges {
   target: string;
 }
 
-export const getGraph = async (
-  services: GraphContextServices,
+interface GetGraphParams {
+  services: GraphContextServices;
   query: {
-    actorIds: string[];
     eventIds: string[];
     spaceId?: string;
     start: string | number;
     end: string | number;
-  }
-): Promise<{
+    esQuery?: EsQuery;
+  };
+  showUnknownTarget: boolean;
+  nodesLimit?: number;
+}
+
+export const getGraph = async ({
+  services: { esClient, logger },
+  query: { eventIds, spaceId = 'default', start, end, esQuery },
+  showUnknownTarget,
+  nodesLimit,
+}: GetGraphParams): Promise<{
   nodes: NodeDataModel[];
   edges: EdgeDataModel[];
 }> => {
-  const { esClient, logger } = services;
-  const { actorIds, eventIds, spaceId = 'default', start, end } = query;
-
-  logger.trace(
-    `Fetching graph for [eventIds: ${eventIds.join(', ')}] [actorIds: ${actorIds.join(
-      ', '
-    )}] in [spaceId: ${spaceId}]`
-  );
-
-  const results = await fetchGraph({ esClient, logger, start, end, eventIds, actorIds });
+  logger.trace(`Fetching graph for [eventIds: ${eventIds.join(', ')}] in [spaceId: ${spaceId}]`);
+
+  const results = await fetchGraph({
+    esClient,
+    showUnknownTarget,
+    logger,
+    start,
+    end,
+    eventIds,
+    esQuery,
+  });
 
   // Convert results into set of nodes and edges
-  const graphContext = parseRecords(logger, results.records);
+  const graphContext = parseRecords(logger, results.records, nodesLimit);
 
   return { nodes: graphContext.nodes, edges: graphContext.edges };
 };
 
 interface ParseContext {
-  nodesMap: Record<string, NodeDataModel>;
-  edgesMap: Record<string, EdgeDataModel>;
-  edgeLabelsNodes: Record<string, string[]>;
-  labelEdges: Record<string, LabelEdges>;
+  readonly nodesLimit?: number;
+  readonly nodesMap: Record<string, NodeDataModel>;
+  readonly edgesMap: Record<string, EdgeDataModel>;
+  readonly edgeLabelsNodes: Record<string, string[]>;
+  readonly labelEdges: Record<string, LabelEdges>;
+  readonly logger: Logger;
 }
 
-const parseRecords = (logger: Logger, records: GraphEdge[]): GraphContext => {
-  const ctx: ParseContext = { nodesMap: {}, edgeLabelsNodes: {}, edgesMap: {}, labelEdges: {} };
+const parseRecords = (logger: Logger, records: GraphEdge[], nodesLimit?: number): GraphContext => {
```

Comment:
should the `parseRecords` also ensure values that we expect to be array are parsed?

the [docs suggests](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-functions-operators.html#esql-agg-values) that string might be returned from `VALUES()` when there's a single value, that could cause error when accessing the `array.includes` method below, i.e `ips.includes(actorId)`

Reply:
good point, 
I'll add tests for these edge cases